src/Desktop.Analyzers/Core/DoNotCatchCorruptedStateExceptions.cs

```diff
@@ -29,132 +30,108 @@ public abstract class DoNotCatchCorruptedStateExceptionsAnalyzer<TLanguageKindEn
-            protected abstract bool IsCatchClause(SyntaxNode node);
-            protected abstract bool IslambdaExpression(SyntaxNode node);
-
-            protected CodeBlockAnalyzer(CompilationSecurityTypes compilationTypes, ISymbol owningSymbol, SyntaxNode codeBlock)
+            private readonly CompilationSecurityTypes _compilationTypes;
+            private readonly Stack<bool> _seenEmptyThrowInCatchClauses = new Stack<bool>();
+
+            public ISet<ICatchClause> CatchAllCatchClausesWithoutEmptyThrow { get; } = new HashSet<ICatchClause>();
+
+            public EmptyThrowInsideCatchAllWalker(CompilationSecurityTypes compilationTypes)
             {
-                _owningSymbol = owningSymbol;
-                _codeBlock = codeBlock;
-                _catchAllCatchClauses = new Dictionary<TCatchClauseSyntax, ISymbol>();
-                TypesOfInterest = compilationTypes;
+                _compilationTypes = compilationTypes;
             }
 
-            public void AnalyzeCatchClause(SyntaxNodeAnalysisContext context)
+            public override void VisitCatch(ICatchClause operation)
             {
-                var catchNode = (TCatchClauseSyntax)context.Node;
-                ISymbol exceptionTypeSymbol = GetExceptionTypeSymbolFromCatchClause(catchNode, context.SemanticModel);
+                _seenEmptyThrowInCatchClauses.Push(false);
 
-                if (IsCatchTypeTooGeneral(exceptionTypeSymbol))
-                {
-                    SyntaxNode parentNode = catchNode.Parent;
-                    while (parentNode != _codeBlock)
-                    {
-                        // for now there doesn't seem to have any way to annotate lambdas with attributes
-                        if (IslambdaExpression(parentNode))
-                        {
-                            return;
-                        }
-                        parentNode = parentNode.Parent;
-                    }
-                    _catchAllCatchClauses[catchNode] = exceptionTypeSymbol;
-                }
-            }
+                Visit(operation.Filter);
+                Visit(operation.Handler);
 
-            public void AnalyzeThrowStatement(SyntaxNodeAnalysisContext context)
-            {
-                var throwNode = (TThrowStatementSyntax)context.Node;
+                bool seenEmptyThrow = _seenEmptyThrowInCatchClauses.Pop();
 
-                // throwNode is a throw statement with no argument, which is not allowed outside of a catch clause
-                if (IsThrowStatementWithNoArgument(throwNode))
+                if (IsCaughtTypeTooGeneral(operation.CaughtType) && !seenEmptyThrow)
                 {
-                    var enlcosingCatchClause = (TCatchClauseSyntax)throwNode.Ancestors().First(IsCatchClause);
-                    _catchAllCatchClauses.Remove(enlcosingCatchClause);
+                    // TODO: Abort in case parent is a lambda.
```

Comment:
I wonder what such a test would look like, because the following does not compile:

```csharp
    class TestClass
    {
        public static void TestMethod()
        {
            try
            {
                FileStream fileStream = new FileStream("name", FileMode.Create);
            }
            catch (Exception)
            {
                Action action = () =>
                {
                    throw; // CS1056: A throw statement with no arguments is not allowed outside of a catch clause
                };
                action();
            }
        }
    }
```

Perhaps this was just an optimization for early bail out? It would explain why no test exists, as it does not influence the outcome.